Implement LegacyLenientSetter WebIDL attribute

tetsuharuohzeki opened the pull request:
#### 1c2466b7b7169d3c4bdf8c7b3deb20173acc3d19
<pre>
Implement LegacyLenientSetter WebIDL attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=249989">https://bugs.webkit.org/show_bug.cgi?id=249989</a>

Reviewed by Chris Dumez and Darin Adler.

This supports <a href="https://webidl.spec.whatwg.org/#LegacyLenientSetter">https://webidl.spec.whatwg.org/#LegacyLenientSetter</a>

* LayoutTests/imported/w3c/web-platform-tests/fullscreen/idlharness.window-expected.txt:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(IsReadonly):
(GenerateAttributeSetterBodyDefinition):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
(WebCore::jsTestObj_legacyLenientSetterTestAttrGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::setJSTestObj_legacyLenientSetterTestAttrSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
* Source/WebCore/bindings/scripts/test/TestObj.idl:

Canonical link: <a href="https://commits.webkit.org/258940@main">https://commits.webkit.org/258940@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/355eb9f362cced0af1226f160dfe521bf69a6f1b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103414 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36378 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112650 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172857 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107367 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3438 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95632 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111666 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109188 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10429 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93517 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38061 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92250 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79792 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5924 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26495 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6098 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3024 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12081 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46004 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6144 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7856 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tetsuharuohzeki commented on Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
```diff
@@ -3289,6 +3292,36 @@ JSC_DEFINE_CUSTOM_SETTER(setJSTestObj_testNullableObjAttr, (JSGlobalObject* lexi
+    UNUSED_PARAM(lexicalGlobalObject);
+    UNUSED_PARAM(thisObject);
+    UNUSED_PARAM(value);
+    RETURN_IF_EXCEPTION(throwScope, false);
```
@Constellation @cdumez 

As my understanding, [`LegacyLenientSetter`](https://webidl.spec.whatwg.org/#LegacyLenientSetter) does nothing.

So then should this `RETURN_IF_EXCEPTION` be  `UNUSED_PARAM(throwScope)` more suitable?

cdumez replied:
I think we can simply `throwScope.assertNoException();` so that we ASSERT() in debug that there is no exception but avoid an unnecessary if check in release.

cdumez reviewed (approved):
LGTM

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5529,6 +5530,17 @@ sub GenerateAttributeSetterBodyDefinition
+        push(@$outputArray, "    UNUSED_PARAM(lexicalGlobalObject);\n");
+        push(@$outputArray, "    UNUSED_PARAM(thisObject);\n");
+        push(@$outputArray, "    UNUSED_PARAM(value);\n");
+        push(@$outputArray, "    throwScope.assertNoException();\n");
```
It would be even better not to define `throwScope` at all in this case, but this is acceptable.

tetsuharuohzeki replied:
I addressed in https://github.com/WebKit/WebKit/pull/8121/commits/355eb9f362cced0af1226f160dfe521bf69a6f1b

tetsuharuohzeki committed:
Implement LegacyLenientSetter WebIDL attribute
https://bugs.webkit.org/show_bug.cgi?id=249989

Reviewed by Chris Dumez and Darin Adler.

This supports https://webidl.spec.whatwg.org/#LegacyLenientSetter

* LayoutTests/imported/w3c/web-platform-tests/fullscreen/idlharness.window-expected.txt:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(IsReadonly):
(GenerateAttributeSetterBodyDefinition):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::JSTestObjDOMConstructor::construct):
(WebCore::jsTestObj_legacyLenientSetterTestAttrGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::setJSTestObj_legacyLenientSetterTestAttrSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
* Source/WebCore/bindings/scripts/test/TestObj.idl:

Canonical link: https://commits.webkit.org/258940@main
